File: Second 100 completed solutions/987. Vertical Order Traversal of a Binary Tree/987. Vertical Order Traversal of a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        q=deque([(root,0,0)])
        nodes=[]
        res=defaultdict(list)
        while q:
            for i in range(len(q)):
                node,row,col=q.popleft()
                nodes.append((col,row,node.val))
                if node.left:
                    q.append((node.left,row+1,col-1))
                if node.right:
                    q.append((node.right,row+1,col+1))
        nodes.sort()
        
        for col, row, value in nodes:
            res[col].append(value)

        return [res[col] for col in res.keys()]

#solution 2 (dfs)
class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        nodes=[]
        res=defaultdict(list)
        def dfs(node,row,col):
            if not node:
                return
            
            nodes.append((col,row,node.val))
            dfs(node.left,row+1,col-1)
            dfs(node.right,row+1,col+1)

        dfs(root,0,0)
        nodes.sort()
        
        for col, row, value in nodes:
            res[col].append(value)

        return list(res.values())
        
        

File: Second 100 completed solutions/987. Vertical Order Traversal of a Binary Tree/987. Vertical Order Traversal of a Binary Tree.cpp

#include <vector>
#include <map>
#include <tuple>
#include <algorithm>

using namespace std;

struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode() : val(0), left(nullptr), right(nullptr) {}
      TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
      TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };
 
class Solution {

private:
    void dfs(TreeNode* node, int row, int col, vector<tuple<int, int, int>>& nodes) {
        if (node == nullptr) {
            return;
        }

        nodes.push_back({col, row, node->val});

        dfs(node->left, row + 1, col - 1, nodes);
        dfs(node->right, row + 1, col + 1, nodes);
    }
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        vector<tuple<int, int, int>> nodes;
        map<int, vector<int>> res;

        dfs(root, 0, 0, nodes);

        sort(nodes.begin(),nodes.end());

        for(auto& [col,row,value] : nodes){
            res[col].push_back(value);
        }

        vector<vector<int>> ans;

        for(auto& [col,values] : res){
            ans.push_back(values);
        }
        return ans;

        
    }
};